File: sonet-client/ios/Sonet/Views/Search/SearchViewModel.swift
import Foundation
import Combine
import SwiftUI

@MainActor
class SearchViewModel: ObservableObject {
    
    // MARK: - Published Properties
    @Published var searchQuery = ""
    @Published var searchResults: SearchResults = SearchResults()
    @Published var trendingTopics: [TrendingTopic] = []
    @Published var recentSearches: [String] = []
    @Published var isSearching = false
    @Published var searchError: String?
    @Published var selectedSearchType: SearchType = .all
    
    // MARK: - Private Properties
    private let grpcClient: SonetGRPCClient
    private var cancellables = Set<AnyCancellable>()
    private var searchTask: Task<Void, Never>?
    
    // MARK: - Initialization
    init(grpcClient: SonetGRPCClient) {
        self.grpcClient = grpcClient
        loadRecentSearches()
        loadTrendingTopics()
        setupSearchDebouncing()
    }
    
    // MARK: - Search Types
    enum SearchType: String, CaseIterable {
        case all = "All"
        case users = "Users"
        case notes = "Notes"
        case hashtags = "Hashtags"
        
        var icon: String {
            switch self {
            case .all: return "magnifyingglass"
            case .users: return "person.2"
            case .notes: return "text.bubble"
            case .hashtags: return "number"
            }
        }
    }
    
    // MARK: - Search Results
    struct SearchResults {
        var users: [UserProfile] = []
        var notes: [Note] = []
        var hashtags: [String] = []
        var hasMoreResults = false
    }
    
    // MARK: - Trending Topic
    struct TrendingTopic: Identifiable {
        let id = UUID()
        let hashtag: String
        let postCount: Int
        let isTrending: Bool
        let category: String?
    }
    
    // MARK: - Public Methods
    func performSearch() {
        guard !searchQuery.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            clearSearchResults()
            return
        }
        
        // Cancel previous search
        searchTask?.cancel()
        
        searchTask = Task {
            await performSearchTask()
        }
    }
    
    func clearSearchResults() {
        searchResults = SearchResults()
        searchError = nil
    }
    
    func addToRecentSearches(_ query: String) {
        let trimmedQuery = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedQuery.isEmpty else { return }
        
        // Remove if already exists
        recentSearches.removeAll { $0 == trimmedQuery }
        
        // Add to beginning
        recentSearches.insert(trimmedQuery, at: 0)
        
        // Keep only last 10 searches
        if recentSearches.count > 10 {
            recentSearches = Array(recentSearches.prefix(10))
        }
        
        saveRecentSearches()
    }
    
    func removeFromRecentSearches(_ query: String) {
        recentSearches.removeAll { $0 == trimmedQuery }
        saveRecentSearches()
    }
    
    func clearRecentSearches() {
        recentSearches.removeAll()
        saveRecentSearches()
    }
    
    // MARK: - Private Methods
    private func setupSearchDebouncing() {
        $searchQuery
            .debounce(for: .milliseconds(300), scheduler: DispatchQueue.main)
            .removeDuplicates()
            .sink { [weak self] query in
                if !query.isEmpty {
                    self?.performSearch()
                } else {
                    self?.clearSearchResults()
                }
            }
            .store(in: &cancellables)
    }
    
    private func performSearchTask() async {
        guard !searchQuery.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return }
        
        await MainActor.run {
            isSearching = true
            searchError = nil
        }
        
        do {
            let query = searchQuery.trimmingCharacters(in: .whitespacesAndNewlines)
            
            // Add to recent searches
            addToRecentSearches(query)
            
            // Perform search based on selected type
            switch selectedSearchType {
            case .all:
                await performComprehensiveSearch(query: query)
            case .users:
                await performUserSearch(query: query)
            case .notes:
                await performNoteSearch(query: query)
            case .hashtags:
                await performHashtagSearch(query: query)
            }
            
        } catch {
            await MainActor.run {
                searchError = error.localizedDescription
            }
        }
        
        await MainActor.run {
            isSearching = false
        }
    }
    
    private func performComprehensiveSearch(query: String) async {
        do {
            async let usersTask = grpcClient.searchUsers(query: query, page: 0, pageSize: 10)
            async let notesTask = grpcClient.searchNotes(query: query, page: 0, pageSize: 10)
            
            let (users, notes) = try await (usersTask, notesTask)
            
            await MainActor.run {
                searchResults.users = users
                searchResults.notes = notes
                searchResults.hashtags = extractHashtags(from: query)
                searchResults.hasMoreResults = users.count >= 10 || notes.count >= 10
            }
            
        } catch {
            await MainActor.run {
                searchError = "Search failed: \(error.localizedDescription)"
            }
        }
    }
    
    private func performUserSearch(query: String) async {
        do {
            let users = try await grpcClient.searchUsers(query: query, page: 0, pageSize: 20)
            
            await MainActor.run {
                searchResults.users = users
                searchResults.notes = []
                searchResults.hashtags = []
                searchResults.hasMoreResults = users.count >= 20
            }
            
        } catch {
            await MainActor.run {
                searchError = "User search failed: \(error.localizedDescription)"
            }
        }
    }
    
    private func performNoteSearch(query: String) async {
        do {
            let notes = try await grpcClient.searchNotes(query: query, page: 0, pageSize: 20)
            
            await MainActor.run {
                searchResults.users = []
                searchResults.notes = notes
                searchResults.hashtags = extractHashtags(from: query)
                searchResults.hasMoreResults = notes.count >= 20
            }
            
        } catch {
            await MainActor.run {
                searchError = "Note search failed: \(error.localizedDescription)"
            }
        }
    }
    
    private func performHashtagSearch(query: String) async {
        // For hashtag search, we'll search notes and extract hashtags
        do {
            let notes = try await grpcClient.searchNotes(query: query, page: 0, pageSize: 50)
            let hashtags = extractHashtagsFromNotes(notes)
            
            await MainActor.run {
                searchResults.users = []
                searchResults.notes = []
                searchResults.hashtags = hashtags
                searchResults.hasMoreResults = false
            }
            
        } catch {
            await MainActor.run {
                searchError = "Hashtag search failed: \(error.localizedDescription)"
            }
        }
    }
    
    private func extractHashtags(from query: String) -> [String] {
        let hashtagPattern = #"#(\w+)"#
        let regex = try? NSRegularExpression(pattern: hashtagPattern)
        let range = NSRange(query.startIndex..<query.endIndex, in: query)
        
        guard let matches = regex?.matches(in: query, range: range) else { return [] }
        
        return matches.compactMap { match in
            guard let range = Range(match.range(at: 1), in: query) else { return nil }
            return String(query[range])
        }
    }
    
    private func extractHashtagsFromNotes(_ notes: [Note]) -> [String] {
        var hashtags: Set<String> = []
        
        for note in notes {
            hashtags.formUnion(note.hashtags)
        }
        
        return Array(hashtags).sorted()
    }
    
    private func loadTrendingTopics() {
        // Load trending topics from gRPC or generate mock data
        trendingTopics = [
            TrendingTopic(hashtag: "#TechNews", postCount: 15420, isTrending: true, category: "Technology"),
            TrendingTopic(hashtag: "#AI", postCount: 12850, isTrending: true, category: "Technology"),
            TrendingTopic(hashtag: "#ClimateAction", postCount: 9870, isTrending: false, category: "Environment"),
            TrendingTopic(hashtag: "#SpaceX", postCount: 7650, isTrending: true, category: "Space"),
            TrendingTopic(hashtag: "#Gaming", postCount: 6540, isTrending: false, category: "Entertainment"),
            TrendingTopic(hashtag: "#Crypto", postCount: 5430, isTrending: false, category: "Finance"),
            TrendingTopic(hashtag: "#Health", postCount: 4320, isTrending: false, category: "Health"),
            TrendingTopic(hashtag: "#Travel", postCount: 3210, isTrending: false, category: "Lifestyle")
        ]
    }
    
    private func loadRecentSearches() {
        recentSearches = UserDefaults.standard.stringArray(forKey: "recentSearches") ?? []
    }
    
    private func saveRecentSearches() {
        UserDefaults.standard.set(recentSearches, forKey: "recentSearches")
    }
}

File: android/app/src/main/java/xyz/sonet/app/viewmodels/SearchViewModel.kt
package xyz.sonet.app.viewmodels

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import xyz.sonet.app.grpc.SonetGRPCClient
import xyz.sonet.app.grpc.SonetConfiguration
import xyz.sonet.app.grpc.proto.*
import java.util.*

class SearchViewModel(application: Application) : AndroidViewModel(application) {
    
    // MARK: - State Flows
    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()
    
    private val _searchResults = MutableStateFlow(SearchResults())
    val searchResults: StateFlow<SearchResults> = _searchResults.asStateFlow()
    
    private val _trendingTopics = MutableStateFlow<List<TrendingTopic>>(emptyList())
    val trendingTopics: StateFlow<List<TrendingTopic>> = _trendingTopics.asStateFlow()
    
    private val _recentSearches = MutableStateFlow<List<String>>(emptyList())
    val recentSearches: StateFlow<List<String>> = _recentSearches.asStateFlow()
    
    private val _isSearching = MutableStateFlow(false)
    val isSearching: StateFlow<Boolean> = _isSearching.asStateFlow()
    
    private val _searchError = MutableStateFlow<String?>(null)
    val searchError: StateFlow<String?> = _searchError.asStateFlow()
    
    private val _selectedSearchType = MutableStateFlow(SearchType.ALL)
    val selectedSearchType: StateFlow<SearchType> = _selectedSearchType.asStateFlow()
    
    // MARK: - Private Properties
    private val grpcClient = SonetGRPCClient(application, SonetConfiguration.development)
    private var searchJob: kotlinx.coroutines.Job? = null
    
    // MARK: - Initialization
    init {
        loadRecentSearches()
        loadTrendingTopics()
    }
    
    // MARK: - Search Types
    enum class SearchType(val title: String, val icon: String) {
        ALL("All", "magnifyingglass"),
        USERS("Users", "person.2"),
        NOTES("Notes", "text.bubble"),
        HASHTAGS("Hashtags", "number")
    }
    
    // MARK: - Data Classes
    data class SearchResults(
        val users: List<UserProfile> = emptyList(),
        val notes: List<Note> = emptyList(),
        val hashtags: List<String> = emptyList(),
        val hasMoreResults: Boolean = false
    )
    
    data class TrendingTopic(
        val id: String = UUID.randomUUID().toString(),
        val hashtag: String,
        val postCount: Int,
        val isTrending: Boolean,
        val category: String?
    )
    
    // MARK: - Public Methods
    fun updateSearchQuery(query: String) {
        _searchQuery.value = query
        performSearch()
    }
    
    fun setSearchType(searchType: SearchType) {
        _selectedSearchType.value = searchType
        if (_searchQuery.value.isNotEmpty()) {
            performSearch()
        }
    }
    
    fun performSearch() {
        val query = _searchQuery.value.trim()
        if (query.isEmpty) {
            clearSearchResults()
            return
        }
        
        // Cancel previous search
        searchJob?.cancel()
        
        searchJob = viewModelScope.launch {
            performSearchTask(query)
        }
    }
    
    fun clearSearchResults() {
        _searchResults.value = SearchResults()
        _searchError.value = null
    }
    
    fun addToRecentSearches(query: String) {
        val trimmedQuery = query.trim()
        if (trimmedQuery.isEmpty) return
        
        val currentSearches = _recentSearches.value.toMutableList()
        
        // Remove if already exists
        currentSearches.remove(trimmedQuery)
        
        // Add to beginning
        currentSearches.add(0, trimmedQuery)
        
        // Keep only last 10 searches
        if (currentSearches.size > 10) {
            currentSearches.subList(10, currentSearches.size).clear()
        }
        
        _recentSearches.value = currentSearches
        saveRecentSearches()
    }
    
    fun removeFromRecentSearches(query: String) {
        val currentSearches = _recentSearches.value.toMutableList()
        currentSearches.remove(query)
        _recentSearches.value = currentSearches
        saveRecentSearches()
    }
    
    fun clearRecentSearches() {
        _recentSearches.value = emptyList()
        saveRecentSearches()
    }
    
    // MARK: - Private Methods
    private suspend fun performSearchTask(query: String) {
        _isSearching.value = true
        _searchError.value = null
        
        try {
            // Add to recent searches
            addToRecentSearches(query)
            
            // Perform search based on selected type
            when (_selectedSearchType.value) {
                SearchType.ALL -> performComprehensiveSearch(query)
                SearchType.USERS -> performUserSearch(query)
                SearchType.NOTES -> performNoteSearch(query)
                SearchType.HASHTAGS -> performHashtagSearch(query)
            }
            
        } catch (error: Exception) {
            _searchError.value = error.message ?: "Search failed"
        } finally {
            _isSearching.value = false
        }
    }
    
    private suspend fun performComprehensiveSearch(query: String) {
        try {
            val users = grpcClient.searchUsers(query, 0, 10)
            val notes = grpcClient.searchNotes(query, 0, 10)
            val hashtags = extractHashtags(query)
            
            _searchResults.value = SearchResults(
                users = users,
                notes = notes,
                hashtags = hashtags,
                hasMoreResults = users.size >= 10 || notes.size >= 10
            )
            
        } catch (error: Exception) {
            _searchError.value = "Search failed: ${error.message}"
        }
    }
    
    private suspend fun performUserSearch(query: String) {
        try {
            val users = grpcClient.searchUsers(query, 0, 20)
            
            _searchResults.value = SearchResults(
                users = users,
                notes = emptyList(),
                hashtags = emptyList(),
                hasMoreResults = users.size >= 20
            )
            
        } catch (error: Exception) {
            _searchError.value = "User search failed: ${error.message}"
        }
    }
    
    private suspend fun performNoteSearch(query: String) {
        try {
            val notes = grpcClient.searchNotes(query, 0, 20)
            val hashtags = extractHashtags(query)
            
            _searchResults.value = SearchResults(
                users = emptyList(),
                notes = notes,
                hashtags = hashtags,
                hasMoreResults = notes.size >= 20
            )
            
        } catch (error: Exception) {
            _searchError.value = "Note search failed: ${error.message}"
        }
    }
    
    private suspend fun performHashtagSearch(query: String) {
        try {
            val notes = grpcClient.searchNotes(query, 0, 50)
            val hashtags = extractHashtagsFromNotes(notes)
            
            _searchResults.value = SearchResults(
                users = emptyList(),
                notes = emptyList(),
                hashtags = hashtags,
                hasMoreResults = false
            )
            
        } catch (error: Exception) {
            _searchError.value = "Hashtag search failed: ${error.message}"
        }
    }
    
    private fun extractHashtags(query: String): List<String> {
        val hashtagPattern = Regex("#(\\w+)")
        return hashtagPattern.findAll(query)
            .map { it.groupValues[1] }
            .toList()
    }
    
    private fun extractHashtagsFromNotes(notes: List<Note>): List<String> {
        val hashtags = mutableSetOf<String>()
        notes.forEach { note ->
            hashtags.addAll(note.hashtagsList)
        }
        return hashtags.sorted()
    }
    
    private fun loadTrendingTopics() {
        // Load trending topics from gRPC or generate mock data
        _trendingTopics.value = listOf(
            TrendingTopic(
                hashtag = "#TechNews",
                postCount = 15420,
                isTrending = true,
                category = "Technology"
            ),
            TrendingTopic(
                hashtag = "#AI",
                postCount = 12850,
                isTrending = true,
                category = "Technology"
            ),
            TrendingTopic(
                hashtag = "#ClimateAction",
                postCount = 9870,
                isTrending = false,
                category = "Environment"
            ),
            TrendingTopic(
                hashtag = "#SpaceX",
                postCount = 7650,
                isTrending = true,
                category = "Space"
            ),
            TrendingTopic(
                hashtag = "#Gaming",
                postCount = 6540,
                isTrending = false,
                category = "Entertainment"
            ),
            TrendingTopic(
                hashtag = "#Crypto",
                postCount = 5430,
                isTrending = false,
                category = "Finance"
            ),
            TrendingTopic(
                hashtag = "#Health",
                postCount = 4320,
                isTrending = false,
                category = "Health"
            ),
            TrendingTopic(
                hashtag = "#Travel",
                postCount = 3210,
                isTrending = false,
                category = "Lifestyle"
            )
        )
    }
    
    private fun loadRecentSearches() {
        // Load from SharedPreferences
        val sharedPrefs = getApplication<Application>().getSharedPreferences("search_prefs", 0)
        val searches = sharedPrefs.getStringSet("recent_searches", emptySet())?.toList() ?: emptyList()
        _recentSearches.value = searches
    }
    
    private fun saveRecentSearches() {
        val sharedPrefs = getApplication<Application>().getSharedPreferences("search_prefs", 0)
        val searches = _recentSearches.value.toSet()
        sharedPrefs.edit().putStringSet("recent_searches", searches).apply()
    }
}